File: contract_zec/backend.py
from fastapi import FastAPI, HTTPException
from nearapi import connect, Account, Contract, KeyPair
from pydantic import BaseModel
import json

app = FastAPI()

# Conectar com a NEAR Testnet
def get_near_connection():
    return connect(
        network_id="testnet",
        key_path="~/.near-credentials/testnet/your-account.json"  # Caminho para a chave de conta
    )

# Criar um modelo de dados para transações
class PaymentRequest(BaseModel):
    to: str
    amount: int

class WithdrawRequest(BaseModel):
    amount: int

@app.post("/pay")
async def pay(payment: PaymentRequest):
    try:
        # Conectar à rede NEAR
        near = get_near_connection()
        sender_account = near.account("payment.giovanna-britto.testnet")  # Subconta
        contract = sender_account.contract("contract_zec.giovanna-britto.testnet")

        # Enviar pagamento
        result = contract.pay(
            sender_account,
            to=payment.to,
            attached_deposit=payment.amount  # Valor em NEAR (em yoctoNEAR)
        )
        
        return {"status": "success", "result": result}

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@app.post("/withdraw")
async def withdraw(withdraw_request: WithdrawRequest):
    try:
        # Conectar à rede NEAR
        near = get_near_connection()
        sender_account = near.account("payment.giovanna-britto.testnet")  # Subconta
        contract = sender_account.contract("contract_zec.giovanna-britto.testnet")

        # Retirar pagamento
        result = contract.withdraw(
            sender_account,
            amount=withdraw_request.amount  # Valor em NEAR (em yoctoNEAR)
        )

        return {"status": "success", "result": result}
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@app.get("/balance/{account_id}")
async def get_balance(account_id: str):
    try:
        # Conectar à rede NEAR
        near = get_near_connection()
        contract = near.account("contract_zec.giovanna-britto.testnet").contract("contract_zec.giovanna-britto.testnet")

        # Consultar saldo
        balance = contract.get_balance(account_id)
        
        return {"account_id": account_id, "balance": balance}

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@app.post("/set_owner")
async def set_owner(new_owner: str):
    try:
        # Conectar à rede NEAR
        near = get_near_connection()
        sender_account = near.account("payment.giovanna-britto.testnet")  # Subconta
        contract = sender_account.contract("contract_zec.giovanna-britto.testnet")

        # Alterar dono do contrato
        result = contract.set_owner(
            sender_account,
            new_owner=new_owner
        )

        return {"status": "success", "result": result}
    
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: contract_zec/backend.js
const express = require("express");
const bodyParser = require("body-parser");
const nearAPI = require("near-api-js");

const app = express();
const port = 3000;

// Configuração do corpo da requisição para JSON
app.use(bodyParser.json());

// Função para conectar à rede NEAR
async function getNearConnection() {
    const config = {
        networkId: "testnet",
        nodeUrl: "https://rpc.testnet.near.org",
        keyStore: new nearAPI.keyStores.InMemoryKeyStore(),
    };

    const near = await nearAPI.connect(config);
    return near;
}

// Rota para pagar
app.post("/pay", async (req, res) => {
    const { to, amount } = req.body;

    try {
        const near = await getNearConnection();
        const account = await near.account("payment.giovanna-britto.testnet");
        const contract = new nearAPI.Contract(account, "contract_zec.giovanna-britto.testnet", {
            changeMethods: ["pay"],
        });

        const result = await contract.pay(
            {
                sender_id: account.accountId,
                to,
                attached_deposit: amount, // valor em yoctoNEAR
            }
        );

        res.json({ status: "success", result });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// Rota para retirar
app.post("/withdraw", async (req, res) => {
    const { amount } = req.body;

    try {
        const near = await getNearConnection();
        const account = await near.account("payment.giovanna-britto.testnet");
        const contract = new nearAPI.Contract(account, "contract_zec.giovanna-britto.testnet", {
            changeMethods: ["withdraw"],
        });

        const result = await contract.withdraw({
            sender_id: account.accountId,
            amount, // valor em yoctoNEAR
        });

        res.json({ status: "success", result });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// Rota para obter o saldo
app.get("/balance/:account_id", async (req, res) => {
    const { account_id } = req.params;

    try {
        const near = await getNearConnection();
        const account = await near.account("contract_zec.giovanna-britto.testnet");
        const contract = new nearAPI.Contract(account, "contract_zec.giovanna-britto.testnet", {
            viewMethods: ["get_balance"],
        });

        const balance = await contract.get_balance({ account_id });

        res.json({ account_id, balance });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// Rota para definir o novo dono do contrato
app.post("/set_owner", async (req, res) => {
    const { new_owner } = req.body;

    try {
        const near = await getNearConnection();
        const account = await near.account("payment.giovanna-britto.testnet");
        const contract = new nearAPI.Contract(account, "contract_zec.giovanna-britto.testnet", {
            changeMethods: ["set_owner"],
        });

        const result = await contract.set_owner({
            sender_id: account.accountId,
            new_owner,
        });

        res.json({ status: "success", result });
    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

// Iniciar o servidor
app.listen(port, () => {
    console.log(`Servidor rodando em http://127.0.0.1:${port}`);
});
